File: src/MNLM denoiser for gravity anomaly/Synthetic_Test_For_MNLM.m
clc;
clear all;
close all;
%% Author: Hanbing Ai
%% How to contact me?
%E-mail:AHB_ECUT@163.com/1724178612@qq.com
%% Date:2022.11.13.
% Read the raw data
DATA_1=load('synthetic.xyz');DATA_2=load('10precent.xyz');
%
[D_1,X,Y]=transform_in(DATA_1,121,121);[D_2,XX,YY]=transform_in(DATA_2,121,121);
%%
%
INPUT=D_2; % Noise-corrupted data
TRUE=D_1; % Noise-free data
%
% figure
% %
% set(gcf,'color',[1 1 1],'units','normalized','position',[0 0 0.4 0.55])
% % 
% imagesc(TRUE-INPUT)
% colormap('hsv')
% colorbar
% set(gca,'FontSize',12)
%% Main program for MNLM to denoise
ds=5;% block size for calculate weight
Ds=[2:2:18];% search block
sigma=0.0001*mean(mean(INPUT));
h=[100:100:1000].*sigma;
%
%%
for i=1:length(Ds)
    for j=1:length(h)
       N_processed(i,j)= sqrt(mean(mean((D_1-NLM_II(INPUT,Ds(i),ds,h(j))).^2)));
    end
end
%
[R,C]=find(N_processed==min(min(N_processed)));
%
OUTPUT_MNLM=NLM_II(INPUT,Ds(R(1)),ds,h(C(1)));
Noise_component_MNLM=INPUT-OUTPUT_MNLM;
Difference_MNLM=D_1-OUTPUT_MNLM;
%
RMS_1=min(min(N_processed));
%
figure
%
set(gcf,'color',[1 1 1],'units','normalized','position',[0 0 0.6 0.7])
% 
subplot(2,2,1)
contourf(X,Y,Noise_component_MNLM)
colormap('jet')
colorbar
title('Noise component','FontSize',12)
set(gca,'FontSize',12)
subplot(2,2,2)
contourf(X,Y,OUTPUT_MNLM)
colorbar
title('Denoised result','FontSize',12)
set(gca,'FontSize',12)
subplot(2,2,3)
contourf(X,Y,Difference_MNLM)
colorbar
title('The difference between the denoised result and the noise-free data','FontSize',12)
set(gca,'FontSize',12)
subplot(2,2,4)
contourf(X,Y,INPUT)
colorbar
title('Noise corrupted result','FontSize',12)
set(gca,'FontSize',12)
% 
% figure
% %
% set(gcf,'color',[1 1 1],'units','normalized','position',[0 0 0.5 0.5])
% % 
% imagesc(Ds,h,N_processed)
% shading interp
% h_c=colorbar('southoutside');
% hold on
% % scatter(Ds(R(1)),h(C(1)),40,'markerfacecolor','r','markeredgecolor','k')
% colormap('bone')
% axis tight
% set(gca,'FontSize',14)
% set(gca,'Ydir','normal')
% h_x=xlabel('Ds');      
% h_y=ylabel('h');   
% set(h_x,'fontsize',14);
% set(h_y,'fontsize',14);
% set(gcf,'unit', 'normalized', 'position',[0.1 0.1 0.5 0.5])
% set(gcf,'color',[1 1 1])
%% SVD
[u,s,v]=svd(D_1);
%
s1=s;s0=s;
s0(abs(s1)<580)=0;
%
OUTPUT_SVD=u*s0*v';
Noise_component_SVD=INPUT-OUTPUT_SVD;
Difference_SVD=D_1-OUTPUT_SVD;
% 
figure
%
set(gcf,'color',[1 1 1],'units','normalized','position',[0 0 0.6 0.7])
% 
subplot(2,2,1)
contourf(X,Y,Noise_component_SVD)
colormap('jet')
colorbar
title('Noise component','FontSize',12)
set(gca,'FontSize',12)
subplot(2,2,2)
contourf(X,Y,OUTPUT_SVD)
colorbar
title('Denoised result','FontSize',12)
set(gca,'FontSize',12)
subplot(2,2,3)
contourf(X,Y,Difference_SVD)
colorbar
title('The difference between the denoised result and the noise-free data','FontSize',12)
set(gca,'FontSize',12)
% 
%% Wavelet
[c,s]=wavedec2(D_1,3,'sym4');  
OUTPUT_wavelet=wrcoef2('a',c,s,'sym4');
Noise_component_wavelet=INPUT-OUTPUT_wavelet;
Difference_wavelet=D_1-OUTPUT_wavelet;
% 
figure
%
set(gcf,'color',[1 1 1],'units','normalized','position',[0 0 0.6 0.7])
% 
subplot(2,2,1)
contourf(X,Y,Noise_component_wavelet)
colormap('jet')
colorbar
title('Noise component','FontSize',12)
set(gca,'FontSize',12)
subplot(2,2,2)
contourf(X,Y,OUTPUT_wavelet)
colorbar
title('Denoised result','FontSize',12)
set(gca,'FontSize',12)
subplot(2,2,3)
contourf(X,Y,Difference_wavelet)
colorbar
title('The difference between the denoised result and the noise-free data','FontSize',12)
set(gca,'FontSize',12)
% 
%% DCT
GRAVITY=dct2(D_1);
[m n]=size(GRAVITY);
%
QC1=zeros(m,n);
for i=1:10
for j=1:10
    QC1(i,j)=GRAVITY(i,j); 
end
end
%
OUTPUT_DCT=idct2(QC1);
Noise_component_DCT=INPUT-OUTPUT_DCT;
Difference_DCT=D_1-OUTPUT_DCT;
% 
figure
%
set(gcf,'color',[1 1 1],'units','normalized','position',[0 0 0.6 0.7])
% 
subplot(2,2,1)
contourf(X,Y,Noise_component_DCT)
colormap('jet')
colorbar
title('Noise component','FontSize',12)
set(gca,'FontSize',12)
subplot(2,2,2)
contourf(X,Y,OUTPUT_DCT)
colorbar
title('Denoised result','FontSize',12)
set(gca,'FontSize',12)
subplot(2,2,3)
contourf(X,Y,Difference_DCT)
colorbar
title('The difference between the denoised result and the noise-free data','FontSize',12)
set(gca,'FontSize',12)
%%
DT_MNLM=transform_out(OUTPUT_MNLM,Y(1,:),X(:,1));DT_X=DT_MNLM(:,2);DT_Y=DT_MNLM(:,1);
DT_SVD=transform_out(OUTPUT_SVD,Y(1,:),X(:,1));
DT_wavelet=transform_out(OUTPUT_wavelet,Y(1,:),X(:,1));
DT_DCT=transform_out(OUTPUT_DCT,Y(1,:),X(:,1));
DT_MNLM(:,1)=DT_X;DT_MNLM(:,2)=DT_Y;
DT_SVD(:,1)=DT_X;DT_SVD(:,2)=DT_Y;
DT_wavelet(:,1)=DT_X;DT_wavelet(:,2)=DT_Y;
DT_DCT(:,1)=DT_X;DT_DCT(:,2)=DT_Y;
%%
DT_Noise_component_MNLM=transform_out(Noise_component_MNLM,Y(1,:),X(:,1));
DT_Noise_component_SVD=transform_out(Noise_component_SVD,Y(1,:),X(:,1));
DT_Noise_component_wavelet=transform_out(Noise_component_wavelet,Y(1,:),X(:,1));
DT_Noise_component_DCT=transform_out(Noise_component_DCT,Y(1,:),X(:,1));
DT_Noise_component_MNLM(:,1)=DT_X;DT_Noise_component_MNLM(:,2)=DT_Y;
DT_Noise_component_SVD(:,1)=DT_X;DT_Noise_component_SVD(:,2)=DT_Y;
DT_Noise_component_wavelet(:,1)=DT_X;DT_Noise_component_wavelet(:,2)=DT_Y;
DT_Noise_component_DCT(:,1)=DT_X;DT_Noise_component_DCT(:,2)=DT_Y;
%%
DT_Difference_MNLM=transform_out(Difference_MNLM,Y(1,:),X(:,1));
DT_Difference_SVD=transform_out(Difference_SVD,Y(1,:),X(:,1));
DT_Difference_wavelet=transform_out(Difference_wavelet,Y(1,:),X(:,1));
DT_Difference_DCT=transform_out(Difference_DCT,Y(1,:),X(:,1));
DT_Difference_MNLM(:,1)=DT_X;DT_Difference_MNLM(:,2)=DT_Y;
DT_Difference_SVD(:,1)=DT_X;DT_Difference_SVD(:,2)=DT_Y;
DT_Difference_wavelet(:,1)=DT_X;DT_Difference_wavelet(:,2)=DT_Y;
DT_Difference_DCT(:,1)=DT_X;DT_Difference_DCT(:,2)=DT_Y;
%%
save denoised_result_SyntheticCase_MNLM.txt -ascii DT_MNLM
save denoised_result_SyntheticCase_SVD.txt -ascii DT_SVD 
save denoised_result_SyntheticCase_wavelet.txt -ascii DT_wavelet
save denoised_result_SyntheticCase_DCT.txt -ascii DT_DCT 
% 
save Noise_component_SyntheticCase_MNLM.txt -ascii DT_Noise_component_MNLM
save Noise_component_SyntheticCase_SVD.txt -ascii DT_Noise_component_SVD 
save Noise_component_SyntheticCase_wavelet.txt -ascii DT_Noise_component_wavelet
save Noise_component_SyntheticCase_DCT.txt -ascii DT_Noise_component_DCT 
% 
save Difference_SyntheticCase_MNLM.txt -ascii DT_Difference_MNLM
save Difference_SyntheticCase_SVD.txt -ascii DT_Difference_SVD 
save Difference_SyntheticCase_wavelet.txt -ascii DT_Difference_wavelet
save Difference_SyntheticCase_DCT.txt -ascii DT_Difference_DCT 

File: src/MNLM_denoiser_for_gravity_anomaly/Synthetic_Test_For_MNLM.py
"""Python equivalent of Synthetic_Test_For_MNLM.m

Author of the original MATLAB code: Hanbing Ai
E-mail: AHB_ECUT@163.com / 1724178612@qq.com
Date: 2022.11.13.

This file reproduces the original MATLAB experiment, including its
unfair baselines (SVD / wavelet / DCT applied to the clean field).
For a fair comparison on the noisy grid, run compare_denoisers.py.

Run:  python3 Synthetic_Test_For_MNLM.py
Outputs default to results/matlab_repro/ so committed MATLAB text
files in this folder are not overwritten.
"""

import argparse
from pathlib import Path

import numpy as np
import pywt
import matplotlib.pyplot as plt
from scipy.fft import dctn, idctn

from NLM_II import NLM_II
from transform_in import transform_in
from transform_out import transform_out

HERE = Path(__file__).resolve().parent


def main(argv=None):
    """Reproduce the original MATLAB synthetic experiment.

    Loads ``synthetic.xyz`` and ``10precent.xyz``, oracle-tunes MNLM
    against the clean field, runs the MATLAB-faithful SVD / wavelet /
    DCT baselines on the *clean* field, writes twelve ASCII result
    files, and optionally shows the four contour figures.

    Parameters
    ----------
    argv : list of str, optional
        Command-line tokens; defaults to ``sys.argv[1:]``.
    """
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--output-dir",
        type=Path,
        default=HERE / "results" / "matlab_repro",
        help="Where to write the twelve ASCII result files",
    )
    parser.add_argument(
        "--show",
        action="store_true",
        help="Display figures",
    )
    args = parser.parse_args(argv)
    out_dir = Path(args.output_dir)
    out_dir.mkdir(parents=True, exist_ok=True)

    # %% Read the raw data
    DATA_1 = np.loadtxt(HERE / "synthetic.xyz")
    DATA_2 = np.loadtxt(HERE / "10precent.xyz")
    #
    D_1, X, Y = transform_in(DATA_1, 121, 121)
    D_2, XX, YY = transform_in(DATA_2, 121, 121)
    # %%
    INPUT = D_2  # Noise-corrupted data
    TRUE = D_1   # Noise-free data

    # %% Main program for MNLM to denoise
    ds = 5  # block size for calculating the weight
    Ds = np.arange(2, 20, 2)  # search block: 2,4,...,18
    sigma = 0.0001 * INPUT.mean()
    h = np.arange(100, 1100, 100) * sigma  # 100*sigma ... 1000*sigma
    #
    N_processed = np.zeros((len(Ds), len(h)))
    for i in range(len(Ds)):
        for j in range(len(h)):
            N_processed[i, j] = np.sqrt(np.mean((D_1 - NLM_II(INPUT, Ds[i], ds, h[j])) ** 2))
    #
    # MATLAB's find() scans column-major and the script keeps R(1),C(1);
    # flatten in Fortran (column-major) order to break ties identically:
    R, C = np.unravel_index(np.argmin(N_processed.flatten(order="F")),
                            N_processed.shape, order="F")
    #
    OUTPUT_MNLM = NLM_II(INPUT, Ds[R], ds, h[C])
    Noise_component_MNLM = INPUT - OUTPUT_MNLM
    Difference_MNLM = D_1 - OUTPUT_MNLM
    #
    RMS_1 = N_processed.min()
    print(f"Best MNLM parameters: Ds={Ds[R]}, h={h[C]:.6g}, RMSE={RMS_1:.6g}")
    #
    fig, axes = plt.subplots(2, 2, figsize=(12, 10), facecolor="white")
    cs = axes[0, 0].contourf(X, Y, Noise_component_MNLM, cmap="jet")
    fig.colorbar(cs, ax=axes[0, 0])
    axes[0, 0].set_title("Noise component", fontsize=12)
    cs = axes[0, 1].contourf(X, Y, OUTPUT_MNLM, cmap="jet")
    fig.colorbar(cs, ax=axes[0, 1])
    axes[0, 1].set_title("Denoised result", fontsize=12)
    cs = axes[1, 0].contourf(X, Y, Difference_MNLM, cmap="jet")
    fig.colorbar(cs, ax=axes[1, 0])
    axes[1, 0].set_title("Difference between denoised result and noise-free data", fontsize=10)
    cs = axes[1, 1].contourf(X, Y, INPUT, cmap="jet")
    fig.colorbar(cs, ax=axes[1, 1])
    axes[1, 1].set_title("Noise corrupted result", fontsize=12)
    fig.suptitle("MNLM")

    # %% SVD
    u, s, vt = np.linalg.svd(D_1)
    #
    s0 = s.copy()
    s0[np.abs(s) < 580] = 0.0
    #
    OUTPUT_SVD = (u * s0) @ vt
    Noise_component_SVD = INPUT - OUTPUT_SVD
    Difference_SVD = D_1 - OUTPUT_SVD
    #
    fig, axes = plt.subplots(2, 2, figsize=(12, 10), facecolor="white")
    cs = axes[0, 0].contourf(X, Y, Noise_component_SVD, cmap="jet")
    fig.colorbar(cs, ax=axes[0, 0])
    axes[0, 0].set_title("Noise component", fontsize=12)
    cs = axes[0, 1].contourf(X, Y, OUTPUT_SVD, cmap="jet")
    fig.colorbar(cs, ax=axes[0, 1])
    axes[0, 1].set_title("Denoised result", fontsize=12)
    cs = axes[1, 0].contourf(X, Y, Difference_SVD, cmap="jet")
    fig.colorbar(cs, ax=axes[1, 0])
    axes[1, 0].set_title("Difference between denoised result and noise-free data", fontsize=10)
    axes[1, 1].axis("off")
    fig.suptitle("SVD")

    # %% Wavelet
    # MATLAB: [c,s] = wavedec2(D_1,3,'sym4'); OUTPUT_wavelet = wrcoef2('a',c,s,'sym4');
    # i.e. keep only the level-3 approximation and reconstruct.
    coeffs = pywt.wavedec2(D_1, "sym4", level=3)
    coeffs_approx_only = [coeffs[0]] + [
        tuple(np.zeros_like(d) for d in details) for details in coeffs[1:]
    ]
    OUTPUT_wavelet = pywt.waverec2(coeffs_approx_only, "sym4")
    OUTPUT_wavelet = OUTPUT_wavelet[: D_1.shape[0], : D_1.shape[1]]
    Noise_component_wavelet = INPUT - OUTPUT_wavelet
    Difference_wavelet = D_1 - OUTPUT_wavelet
    #
    fig, axes = plt.subplots(2, 2, figsize=(12, 10), facecolor="white")
    cs = axes[0, 0].contourf(X, Y, Noise_component_wavelet, cmap="jet")
    fig.colorbar(cs, ax=axes[0, 0])
    axes[0, 0].set_title("Noise component", fontsize=12)
    cs = axes[0, 1].contourf(X, Y, OUTPUT_wavelet, cmap="jet")
    fig.colorbar(cs, ax=axes[0, 1])
    axes[0, 1].set_title("Denoised result", fontsize=12)
    cs = axes[1, 0].contourf(X, Y, Difference_wavelet, cmap="jet")
    fig.colorbar(cs, ax=axes[1, 0])
    axes[1, 0].set_title("Difference between denoised result and noise-free data", fontsize=10)
    axes[1, 1].axis("off")
    fig.suptitle("Wavelet")

    # %% DCT
    GRAVITY = dctn(D_1, type=2, norm="ortho")
    m, n = GRAVITY.shape
    #
    QC1 = np.zeros((m, n))
    QC1[:10, :10] = GRAVITY[:10, :10]
    #
    OUTPUT_DCT = idctn(QC1, type=2, norm="ortho")
    Noise_component_DCT = INPUT - OUTPUT_DCT
    Difference_DCT = D_1 - OUTPUT_DCT
    #
    fig, axes = plt.subplots(2, 2, figsize=(12, 10), facecolor="white")
    cs = axes[0, 0].contourf(X, Y, Noise_component_DCT, cmap="jet")
    fig.colorbar(cs, ax=axes[0, 0])
    axes[0, 0].set_title("Noise component", fontsize=12)
    cs = axes[0, 1].contourf(X, Y, OUTPUT_DCT, cmap="jet")
    fig.colorbar(cs, ax=axes[0, 1])
    axes[0, 1].set_title("Denoised result", fontsize=12)
    cs = axes[1, 0].contourf(X, Y, Difference_DCT, cmap="jet")
    fig.colorbar(cs, ax=axes[1, 0])
    axes[1, 0].set_title("Difference between denoised result and noise-free data", fontsize=10)
    axes[1, 1].axis("off")
    fig.suptitle("DCT")

    # %% Convert grids back to XYZ lists (with the same column swap as the MATLAB script)
    def _to_xyz(grid):
        """Flatten a grid to XYZ and swap X/Y to match the MATLAB script."""
        cc = transform_out(grid, Y[0, :], X[:, 0])
        # MATLAB swaps columns 1 and 2 after transform_out:
        return np.column_stack((cc[:, 1], cc[:, 0], cc[:, 2]))

    DT_MNLM = _to_xyz(OUTPUT_MNLM)
    DT_SVD = _to_xyz(OUTPUT_SVD)
    DT_wavelet = _to_xyz(OUTPUT_wavelet)
    DT_DCT = _to_xyz(OUTPUT_DCT)
    # %%
    DT_Noise_component_MNLM = _to_xyz(Noise_component_MNLM)
    DT_Noise_component_SVD = _to_xyz(Noise_component_SVD)
    DT_Noise_component_wavelet = _to_xyz(Noise_component_wavelet)
    DT_Noise_component_DCT = _to_xyz(Noise_component_DCT)
    # %%
    DT_Difference_MNLM = _to_xyz(Difference_MNLM)
    DT_Difference_SVD = _to_xyz(Difference_SVD)
    DT_Difference_wavelet = _to_xyz(Difference_wavelet)
    DT_Difference_DCT = _to_xyz(Difference_DCT)

    # %% Save results (equivalent of MATLAB's `save -ascii`)
    _FMT = "%16.7e"
    np.savetxt(out_dir / "denoised_result_SyntheticCase_MNLM.txt", DT_MNLM, fmt=_FMT)
    np.savetxt(out_dir / "denoised_result_SyntheticCase_SVD.txt", DT_SVD, fmt=_FMT)
    np.savetxt(out_dir / "denoised_result_SyntheticCase_wavelet.txt", DT_wavelet, fmt=_FMT)
    np.savetxt(out_dir / "denoised_result_SyntheticCase_DCT.txt", DT_DCT, fmt=_FMT)
    #
    np.savetxt(out_dir / "Noise_component_SyntheticCase_MNLM.txt", DT_Noise_component_MNLM, fmt=_FMT)
    np.savetxt(out_dir / "Noise_component_SyntheticCase_SVD.txt", DT_Noise_component_SVD, fmt=_FMT)
    np.savetxt(out_dir / "Noise_component_SyntheticCase_wavelet.txt", DT_Noise_component_wavelet, fmt=_FMT)
    np.savetxt(out_dir / "Noise_component_SyntheticCase_DCT.txt", DT_Noise_component_DCT, fmt=_FMT)
    #
    np.savetxt(out_dir / "Difference_SyntheticCase_MNLM.txt", DT_Difference_MNLM, fmt=_FMT)
    np.savetxt(out_dir / "Difference_SyntheticCase_SVD.txt", DT_Difference_SVD, fmt=_FMT)
    np.savetxt(out_dir / "Difference_SyntheticCase_wavelet.txt", DT_Difference_wavelet, fmt=_FMT)
    np.savetxt(out_dir / "Difference_SyntheticCase_DCT.txt", DT_Difference_DCT, fmt=_FMT)

    print(f"Wrote MATLAB-repro outputs to {out_dir}")
    if args.show:
        plt.show()
    else:
        plt.close("all")


if __name__ == "__main__":
    main()
